Deprecated Flag for Uncrawlable Publishers

addie9800 opened the pull request:
This PR adds a flag for uncrawlable publishers. The Publisher Coverage Script is updated to skip publishers marked as deprecated. Also, deprecated publishers are marked as such by strike-through in the supported_publishers.md

MaxDall commented on src/fundus/scraping/crawler.py:
```diff
@@ -204,7 +204,10 @@ def build_extraction_filter() -> Optional[ExtractionFilter]:
                     )
                 else:
-                    fitting_publishers.append(publisher)
+                    if not publisher.deprecated:
+                        fitting_publishers.append(publisher)
+                    else:
+                        logger.warning(f"Skipping deprecated publisher: {publisher.publisher_name}")
```
I would either introduce a parameter to include deprecated publisher, since it is still possible to crawl them depending on IP address and User Agent, or don't skip them during crawling.

addie9800 replied:
Good idea, I added that.

MaxDall reviewed (changes requested):
Thanks for adding! Looks awesome 👍

MaxDall commented on src/fundus/scraping/crawler.py:
```diff
@@ -150,6 +150,7 @@ def crawl(
         only_complete: Union[bool, ExtractionFilter] = Requires("title", "body", "publishing_date"),
         url_filter: Optional[URLFilter] = None,
         only_unique: bool = True,
+        ignore_deprecated: bool = False,
```
I would prefer to set the default to `True`, it doesn't hurt if the deprecated publishers cannot be crawled, but if they can and are excluded by default most won't notice. As an example, crawling Reuters works perfectly fine for me.

MaxDall commented on src/fundus/scraping/crawler.py:
```diff
@@ -209,7 +212,10 @@ def build_extraction_filter() -> Optional[ExtractionFilter]:
                     )
                 else:
-                    fitting_publishers.append(publisher)
+                    if (not publisher.deprecated) or ignore_deprecated:
+                        fitting_publishers.append(publisher)
+                    else:
+                        logger.warning(f"Skipping deprecated publisher: {publisher.publisher_name}")
```
Here's a logic bug included. If `only_complete` is set to `False`, the `ignore_deprecated` flag doesn't work anymore, because it is the `if isinstance(extraction_filter, Requires)` block.

addie9800 replied:
Thanks for cathing this

MaxDall commented:
@addie9800 I moved the `ignore_deprecated` flag from the `crawl` method to `Crawler`. The paradigm of `crawl` is to only include parameters affecting all crawlers (forward, backward). Since the depreciation flag does not serve any purpose regarding backward crawling, I decided to move it to the crawler. What do you think about that?

addie9800 commented:
> @addie9800 I moved the `ignore_deprecated` flag from the `crawl` method to `Crawler`. The paradigm of `crawl` is to only include parameters affecting all crawlers (forward, backward). Since the depreciation flag does not serve any purpose regarding backward crawling, I decided to move it to the crawler. What do you think about that?

I think it's a good idea. I also updated the docs a bit. Maybe you can also say what you think about that

MaxDall commented:
@addie9800 thanks for updating the documentation. Unfortunately the latest documentation commit breaks a lot of links. Also tutorial 4 now seems completely overwhelming and out of focus while tutorial 5 looks like it was cannibalized. The purpose of tutorial 4 is to enable the user to filter for specific articles. `search` or `deprecatrd` isn't as nearly as important here. The latter ones are mostly for convenience. I would let tutorial 4 as it was and add the deprecation mechanica to the former tutorial 5.

addie9800 commented:
Oh, and I thought I had caught the broken links. Well, I undid the previous changes and just appended it to section 5 of the tutorial

MaxDall reviewed (approved):
Thanks a lot! Looks really great :)
